syntax = "proto3";

option go_package = "github.com/mingkaic/tenncor/pbm";

package tenncor;

message ByteArr { bytes data = 1; }
message DoubleArr { repeated double data = 1; }
message FloatArr { repeated float data = 1; }
message Int32Arr { repeated int32 data = 1; }
message Int64Arr { repeated int64 data = 1; }
message Uint32Arr { repeated uint32 data = 1; }
message Uint64Arr { repeated uint64 data = 1; }

message Source
{
	bytes shape = 1;
    oneof data {
        ByteArr sbyte_arrs = 2;
        ByteArr ubyte_arrs = 3;
        DoubleArr double_arrs = 4;
        FloatArr float_arrs = 5;
        Int32Arr sshort_arrs = 6;
        Int32Arr sint_arrs = 7;
        Int64Arr slong_arrs = 8;
        Uint32Arr ushort_arrs = 9;
        Uint32Arr uint_arrs = 10;
        Uint64Arr ulong_arrs = 11;
    }
}

message Functor
{
    uint32 opcode = 1;
    // indices of args in graph
    repeated uint32 args = 2;
    repeated uint32 meta = 3;
}

message Node
{
    oneof detail {
        Source source = 1;
        Functor functor = 2;
    }
}

message Graph
{
	string id = 1;
	repeated Node nodes = 2;
}
